services/automerge/automerge.go

```
@@ -0,0 +1,248 @@
+	lastCommitStatus, err := pull_service.GetPullRequestCommitStatusState(ctx, pull)
+	if err != nil {
+		return false, err
+	}
+
+	// we don't need to schedule
+	if lastCommitStatus.IsSuccess() {
+		return false, nil
+	}
+
+	return true, pull_model.ScheduleAutoMerge(ctx, doer, pull.ID, style, message)
+}
+
+// MergeScheduledPullRequest merges a previously scheduled pull request when all checks succeeded
+func MergeScheduledPullRequest(ctx context.Context, sha string, repo *repo_model.Repository) error {
+	pulls, err := getPullRequestsByHeadSHA(ctx, sha, repo, func(pr *models.PullRequest) bool {
+		return !pr.HasMerged && pr.CanAutoMerge()
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, pr := range pulls {
+		add2Queue(pr, sha)
+	}
+
+	return nil
+}
+
+func getPullRequestsByHeadSHA(ctx context.Context, sha string, repo *repo_model.Repository, filter func(*models.PullRequest) bool) (map[int64]*models.PullRequest, error) {
+	gitRepo, err := git.OpenRepository(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
+	defer gitRepo.Close()
+
+	refs, err := gitRepo.GetRefsBySha(sha, "")
+	if err != nil {
+		return nil, err
+	}
+
+	pulls := make(map[int64]*models.PullRequest)
+
+	for _, ref := range refs {
+		// If the branch starts with "pull/*" we know we're dealing with a fork.
+		// In that case, head and base branch are not in the same repo and we need to do some extra work
+		// to get the pull request for this branch.
+		// Each pull branch starts with refs/pull/ we then go from there to find the index of the pr and then
+		// use that to get the pr.
+		if strings.HasPrefix(ref, git.PullPrefix) {
+			parts := strings.Split(ref[len(git.PullPrefix):], "/")
+
+			// e.g. 'refs/pull/1/head' would be []string{"1", "head"}
+			if len(parts) != 2 {
+				continue
+			}
+
+			prIndex, err := strconv.ParseInt(parts[0], 10, 64)
+			if err != nil {
+				return nil, err
+			}
```

Should the error be ignored if it's not a pull number? Like below, if PR not exists, just ignore.